src/function/cast/variant/to_variant.cpp

```diff
@@ -0,0 +1,997 @@
+		keys_entry.length = dictionary_size;
+		keys_entry.offset = dictionary_size * i;
+
+		//! children
+		children_entry.length = children_sizes[i];
+		children_entry.offset = children_offset;
+		children_offset += children_entry.length;
+
+		//! values
+		values_entry.length = values_sizes[i];
+		values_entry.offset = values_offset;
+		values_offset += values_entry.length;
+
+		//! value
+		blob_data[i] = StringVector::EmptyString(blob, blob_sizes[i]);
+	}
+	auto keys_list_size = dictionary_size * offsets.size();
+
+	//! Reserve for the children of the lists
+	ListVector::Reserve(keys, keys_list_size);
+	ListVector::Reserve(children, children_offset);
+	ListVector::Reserve(values, values_offset);
+
+	//! Set list sizes
+	ListVector::SetListSize(keys, keys_list_size);
+	ListVector::SetListSize(children, children_offset);
+	ListVector::SetListSize(values, values_offset);
+
+	keys_selvec.Initialize(keys_list_size);
+	//! NOTE: we initialize these to 0 because some rows will not set them, if the row is NULL
+	//! To simplify the implementation we just allocate 'dictionary.Size()' keys for each row
+	for (idx_t i = 0; i < keys_list_size; i++) {
+		keys_selvec.set_index(i, 0);
+	}
+	selvec_size = keys_list_size;
+}
+
+static bool CastToVARIANT(Vector &source, Vector &result, idx_t count, CastParameters &parameters) {
+	DataChunk offsets;
+	offsets.Initialize(Allocator::DefaultAllocator(),
+	                   {LogicalType::UINTEGER, LogicalType::UINTEGER, LogicalType::UINTEGER, LogicalType::UINTEGER},
+	                   count);
+	offsets.SetCardinality(count);
+	auto &keys = VariantVector::GetKeys(result);
+
+	//! Initialize the dictionary
+	StringDictionary dictionary;
+	{
+		TypeVisitor::Contains(source.GetType(), [&dictionary, &keys](const LogicalType &type) {
+			if (type.id() == LogicalTypeId::STRUCT) {
+				auto &children = StructType::GetChildTypes(type);
+				for (auto &child : children) {
+					string_t child_name_str(child.first.c_str(), NumericCast<uint32_t>(child.first.size()));
+					dictionary.AddString(keys, child_name_str);
+				}
+			} else if (type.id() == LogicalTypeId::MAP) {
+				dictionary.AddString(keys, string_t("key", 3));
+				dictionary.AddString(keys, string_t("value", 5));
+			}
+			return false;
+		});
+	}
+	SelectionVector keys_selvec;
+
+	{
+		VariantVectorData result_data(result);
+		//! First pass - collect sizes/offsets
+		InitializeOffsets(offsets, count);
+		ConvertToVariant<false>(source, result_data, offsets, count, nullptr, keys_selvec, dictionary, nullptr);
+	}
+
+	//! This resizes the lists, invalidating the "GetData" results stored in VariantVectorData
+	idx_t keys_selvec_size;
+	InitializeVariants(offsets, result, keys_selvec, keys_selvec_size, dictionary);
+
+	{
+		VariantVectorData result_data(result);
+		//! Second pass - actually construct the variants
+		InitializeOffsets(offsets, count);
+		ConvertToVariant<true>(source, result_data, offsets, count, nullptr, keys_selvec, dictionary, nullptr);
+	}
+
```

Comment:
Working on a follow-up PR to convert parquet VARIANT to DuckDB VARIANT rather than JSON, I found a potential problem here:
If the `data` of the VARIANT ends up being smaller than `INLINE_BYTES`, the remaining inlined bytes are left uninitialized, this likely needs a `SetSizeAndFinalize` call.

Maybe the `keys` need the same treatment, though I didn't run into that being a problem yet